File: frameworks/java-selenium-rest-assured/pages/StorePage.java
package pages;

import helpers.Steps;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.Rectangle;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;

public final class StorePage extends BasePage {
    public StorePage(WebDriver driver) { super(driver); }

    public void openStore() {
        Steps.step(Steps.Actions.STORE_OPEN, () -> open("/store"));
    }

    public void searchProduct(int id, String name) {
        Steps.step(Steps.Actions.STORE_SEARCH, () -> {
            WebElement input = visible("store-search");
            input.clear();
            input.sendKeys(name);
            visible("product-add-to-cart-" + id);
        });
    }

    public void openProduct(int id, String name) {
        openStore();
        searchProduct(id, name);
    }

    public Map<String, Integer> readInventory(int id) {
        return Steps.stepReturning(Steps.Actions.STORE_READ_INVENTORY, () -> {
            // Retry the click until the modal opens.  On mobile emulation the first
            // click may land during a brief post-navigation card transition and be
            // absorbed without opening the modal; scrolling the button to the centre
            // of the viewport before each attempt mirrors Playwright's actionability
            // checks and allows the transition to finish between polls.
            wait.until(d -> {
                try {
                    WebElement btn = ExpectedConditions.elementToBeClickable(
                            testId("product-stock-info-" + id)).apply(d);
                    if (btn == null) return null;
                    ((JavascriptExecutor) d).executeScript(
                            "arguments[0].scrollIntoView({block:'center'})", btn);
                    btn.click();
                    return ExpectedConditions.visibilityOfElementLocated(
                            testId("stock-info-modal")).apply(d);
                } catch (org.openqa.selenium.StaleElementReferenceException e) {
                    return null;
                }
            });
            String text = visible("stock-info-modal").getText();
            Map<String, Integer> values = Map.of(
                    "stock", numberAfter(text, "Total Stock"),
                    "reserved", numberAfter(text, "Reserved in Your Cart"),
                    "available", numberAfter(text, "Available to Add"));
            visible("stock-info-close").click();
            return values;
        });
    }

    private int numberAfter(String text, String label) {
        Matcher match = Pattern.compile(Pattern.quote(label) + ":\\s*(\\d+)",
                Pattern.CASE_INSENSITIVE).matcher(text);
        if (!match.find()) throw new AssertionError("Inventory value not found for " + label);
        return Integer.parseInt(match.group(1));
    }

    public void addToCart(int id) {
        Steps.step(Steps.Actions.STORE_ADD_TO_CART,
                () -> visible("product-add-to-cart-" + id).click());
    }

    public void expectReserved(int id, int expectedReserved) {
        Steps.step(Steps.Actions.STORE_VERIFY_RESERVED, () -> {
            // Use textContent (all DOM text regardless of CSS visibility) so that the check
            // matches Playwright's toContainText behaviour on mobile layouts where the
            // reserved-count node may be CSS-hidden until the cart API responds.
            wait.until(d -> {
                try {
                    String text = (String) ((JavascriptExecutor) d)
                            .executeScript("return arguments[0].textContent",
                                    d.findElement(testId("product-stock-info-" + id))
                                     .findElement(org.openqa.selenium.By.xpath("..")));
                    return text != null && text.contains("Reserved: " + expectedReserved);
                } catch (org.openqa.selenium.StaleElementReferenceException e) {
                    return false;
                }
            });
        });
    }

    public void openOrderHistory(int productId) {
        Steps.step(Steps.Actions.STORE_OPEN_ORDER_HISTORY, () -> {
            visible("product-order-history-" + productId).click();
            visible("order-history-modal");
        });
    }

    public void expectOrderRow(int orderId, int quantity, String status) {
        Steps.step(Steps.Actions.STORE_VERIFY_ORDER_ROW, () -> {
            WebElement row = visible("order-history-row-" + orderId);
            String text = row.getText().toLowerCase();
            if (!text.contains(String.valueOf(quantity))) {
                throw new AssertionError("Order history row does not show quantity " + quantity);
            }
            if (!text.contains(status.toLowerCase())) {
                throw new AssertionError("Order history row does not show status " + status);
            }
        });
    }

    public void expectMobileModalContract() {
        Steps.step(Steps.Actions.STORE_VERIFY_MOBILE_MODAL, () -> {
            WebElement modal = visible("order-history-modal");
            Rectangle rect = modal.getRect();
            JavascriptExecutor js = (JavascriptExecutor) driver;
            long viewportWidth = ((Number) js.executeScript("return window.innerWidth")).longValue();
            long viewportHeight = ((Number) js.executeScript("return window.innerHeight")).longValue();
            if (rect.getX() < 0
                    || rect.getX() + rect.getWidth() > viewportWidth
                    || rect.getHeight() > viewportHeight) {
                throw new AssertionError("Order history modal is clipped on mobile");
            }
            Boolean backdropOwnsHitTarget = (Boolean) js.executeScript(
                    "const modal=arguments[0], backdrop=modal.parentElement;"
                            + "return getComputedStyle(backdrop).position==='fixed'"
                            + " && getComputedStyle(modal).overflowY==='auto'"
                            + " && backdrop.contains(document.elementFromPoint(1,1));",
                    modal);
            if (!Boolean.TRUE.equals(backdropOwnsHitTarget)) {
                throw new AssertionError(
                        "Backdrop does not block the page or the panel cannot scroll");
            }
            visible("order-history-close").click();
        });
    }
}

File: frameworks/typescript-cypress/pages/store_page.ts
import { ACTIONS, step } from '../helpers/steps';

export type Inventory = { stock: number; reserved: number; available: number };

export class StorePage {
  open(): void {
    step(ACTIONS.storeOpen, () => {
      cy.visit('/store');
    });
  }

  searchProduct(productId: number, productName: string): void {
    step(ACTIONS.storeSearch, () => {
      cy.get('[data-testid="store-search"]').clear().type(productName);
      cy.get(`[data-testid="product-add-to-cart-${productId}"]`).should('be.visible');
    });
  }

  openProduct(productId: number, productName: string): void {
    this.open();
    this.searchProduct(productId, productName);
  }

  readInventory(productId: number): Cypress.Chainable<Inventory> {
    return step(ACTIONS.storeReadInventory, () => {
      cy.get(`[data-testid="product-stock-info-${productId}"]`).click();
      return cy
        .get('[data-testid="stock-info-modal"]')
        .should('be.visible')
        .invoke('text')
        .then((text) => {
          const read = (label: string): number => {
            const match = text.match(new RegExp(`${label}:\\s*(\\d+)`, 'i'));
            if (!match) throw new Error(`Inventory value not found for ${label}`);
            return Number(match[1]);
          };
          const inventory: Inventory = {
            stock: read('Total Stock'),
            reserved: read('Reserved in Your Cart'),
            available: read('Available to Add'),
          };
          return cy.get('[data-testid="stock-info-close"]').click().then(() => inventory);
        });
    });
  }

  addToCart(productId: number): void {
    step(ACTIONS.storeAddToCart, () => {
      cy.intercept('POST', '/api/cart').as('addToCart');
      cy.get(`[data-testid="product-add-to-cart-${productId}"]`).click();
      cy.wait('@addToCart').its('response.statusCode').should('be.oneOf', [200, 201]);
    });
  }

  expectReserved(productId: number, expectedReserved: number): void {
    step(ACTIONS.storeVerifyReserved, () => {
      cy.get(`[data-testid="product-stock-info-${productId}"]`)
        .parent()
        .should('contain.text', `Reserved: ${expectedReserved}`);
    });
  }

  openOrderHistory(productId: number): void {
    step(ACTIONS.storeOpenOrderHistory, () => {
      cy.get(`[data-testid="product-order-history-${productId}"]`).click();
      cy.get('[data-testid="order-history-modal"]').should('be.visible');
    });
  }

  expectOrderRow(orderId: number, quantity: number, status: string): void {
    step(ACTIONS.storeVerifyOrderRow, () => {
      cy.get(`[data-testid="order-history-row-${orderId}"]`)
        .scrollIntoView()
        .should('be.visible')
        .invoke('text')
        .should('include', String(quantity))
        .and('include', status);
    });
  }

  expectMobileModalContract(): void {
    step(ACTIONS.storeVerifyMobileModal, () => {
      cy.get('[data-testid="order-history-modal"]').then(($modal) => {
        const modal = $modal[0];
        const box = modal.getBoundingClientRect();
        expect(box.left, 'modal left edge').to.be.at.least(0);
        expect(box.top, 'modal top edge').to.be.at.least(0);
        expect(box.right, 'modal right edge').to.be.at.most(Cypress.config('viewportWidth'));
        expect(box.height, 'modal height').to.be.at.most(Cypress.config('viewportHeight'));

        const backdrop = modal.parentElement;
        expect(backdrop, 'modal backdrop').to.not.be.null;
        expect(getComputedStyle(backdrop!).position, 'backdrop position').to.eq('fixed');
        expect(getComputedStyle(modal).overflowY, 'modal overflow-y').to.eq('auto');
        const hitTarget = modal.ownerDocument.elementFromPoint(1, 1);
        expect(
          hitTarget !== null && backdrop!.contains(hitTarget),
          'backdrop owns the top-left hit target',
        ).to.eq(true);
      });
      cy.get('[data-testid="order-history-close"]').should('be.visible').click();
      cy.get('[data-testid="order-history-modal"]').should('not.exist');
    });
  }
}
